packages/aws-appsync/src/client.js

```diff
@@ -28,51 +84,47 @@ class AWSAppSyncClient extends ApolloClient {
 
     /**
      *
-     * @param {string} url
+     * @param {object} appSyncOptions
      * @param {ApolloClientOptions<InMemoryCache>} options
      */
-    constructor({ url, region, auth, conflictResolver, complexObjectsCredentials, disableOffline = false }, options) {
-        if (!url || !region || !auth) {
+    constructor({
+        url,
+        region,
+        auth,
+        conflictResolver,
+        complexObjectsCredentials,
+        cacheOptions,
+        disableOffline = false
+    } = {}, options = {}) {
+        const { cache: customCache, link: customLink } = options;
+
+        if (!customLink && (!url || !region || !auth)) {
             throw new Error(
                 'In order to initialize AWSAppSyncClient, you must specify url, region and auth properties on the config object.'
             );
         }
 
-        let res;
-        this.hydratedPromise = new Promise((resolve, reject) => {
-            res = resolve;
-        });
+        let resolveClient;
+
+        const store = disableOffline ? null : createStore(() => this, () => resolveClient(this), conflictResolver);
+        const cache = disableOffline ? (customCache || new InMemoryCache(cacheOptions)) : new OfflineCache(store, cacheOptions);
+
+        const waitForRehydrationLink = new ApolloLink((op, forward) => {
+            let handle = null;
 
-        const store = disableOffline ? null : createStore(
-            this,
-            () => {
-                store.dispatch({ type: 'REHYDRATE_STORE' });
-                res(this);
-            },
-            conflictResolver,
-        );
-        const cache = disableOffline ? new InMemoryCache() : new OfflineCache(store);
-
-        const passthrough = (op, forward) => (forward ? forward(op) : Observable.of());
-        let link = ApolloLink.from([
-            disableOffline ? passthrough : new OfflineLink(store),
-            new ComplexObjectLink(complexObjectsCredentials),
-            new AuthLink({ url, region, auth }),
-            ApolloLink.split(
-                operation => {
-                    const { query } = operation;
-                    const { kind, operation: graphqlOperation } = getMainDefinition(query);
-                    const isSubscription = kind === 'OperationDefinition' && graphqlOperation === 'subscription';
-
-                    return isSubscription;
-                },
-                ApolloLink.from([
-                    new NonTerminatingHttpLink('subsInfo', { uri: url }, true),
-                    new SubscriptionHandshakeLink('subsInfo'),
-                ]),
-                new HttpLink({ uri: url }),
-            ),
-        ]);
+            return new Observable(observer => {
+                this.hydratedPromise.then(() => {
+                    handle = passthrough(op, forward).subscribe(observer);
+                }).catch(observer.error);
+
+                return () => {
```

Comment:
this return will run before waiting hydratedPromise?

Reply:
Yes, an `Observable` *"Returns a cleanup function which will cancel the event stream"*. For reference, see: https://github.com/tc39/proposal-observable#example-observing-keyboard-events